### ISSUE TITLE
dcr/spv: auto-discover redemption when blockchain forks (without restart)

### ISSUE THREAD

**Issue by norwnd** [decred/dcrdex#2076]
Stumbled upon interesting case, using `testnets`. I was testing refunds, initiated DCR-ETH trade with myself, waited until Maker/Taker swaps have been broadcasted, and shut down `dexc`, waited ~30 mins and started `dexc`. Maker published redeem txn, it soon got mined, but Maker couldn't notify server about his redeem (because server doesn't keep track of revoked matches ?)

<img width="200" alt="image" src="https://user-images.githubusercontent.com/112318969/214648006-65527ded-d860-4636-9bd6-5fb6270c2da7.png"><img width="200" alt="image" src="https://user-images.githubusercontent.com/112318969/214649003-5448b892-9763-4cf7-8cc6-03cbee3eb8d9.png">

but 2 hours later Taker still can't find Maker redeem:

<img width="200" alt="image" src="https://user-images.githubusercontent.com/112318969/214648954-bf390b79-42c7-4ea5-8b08-a1370fc25bf9.png">

[Redemption txn](https://testnet.dcrdata.org/tx/57c6505bc9e3342a1b8ea199b978e52b2e06bde89dab0d8d0f51de51f134375c/out/0) got mined in 1066748 (0000000022d52e3e33287ebce49fa96a7d5e3adea5dce02c50b144424ac32a94) block (on main-chain).

digging through the logs, here is what I think happened:
```
2023-01-25 15:54:27.095 [DBG] CORE[dcr]: finding redemptions for 1 contracts in blocks 1066738 - 1066748
2023-01-25 15:54:27.096 [DBG] CORE[dcr]: 0 redemptions found, 0 canceled out of 1 contracts in blocks 1066738 to 1066748
```
here ^ my DCR SPV wallet was looking at 1066748 on side-chain (see more on this below; probably on side-chain txn hasn't been mined in 1066748 yet, while it has been on main-chain), 
then search continues, **never re-checking 1066748 block** (which **contains** redemption txn on main-chain)
```
2023-01-25 16:00:59.881 [DBG] CORE[dcr]: finding redemptions for 1 contracts in blocks 1066749 - 1066749
2023-01-25 16:00:59.882 [DBG] CORE[dcr]: 0 redemptions found, 0 canceled out of 1 contracts in blocks 1066749 to 1066749
```
and here is another relevant bit from my logs (looks like it was searching side-chain block 1066748, not main-chain one):
```
2023-01-25 15:54:19.564 [INF] CORE[dcr]: Output 466fc2e455c4caf6c77046c21b98f82a036f5f516a18ec32ad99b655783f9246:0 NOT yet found; now searching with block filters.
2023-01-25 15:54:19.564 [DBG] CORE[dcr]: Searching for tx 466fc2e455c4caf6c77046c21b98f82a036f5f516a18ec32ad99b655783f9246 in blocks between best block 1066747 (00000000a96f16971b1a09f4d1e634d1f196f45ce32f4c74e1bf60c7308650b2) and the block just before 2023-01-25 15:23:45 +0000 UTC.
2023-01-25 15:54:19.626 [DBG] CORE[dcr]: tip change: 1066747 (00000000a96f16971b1a09f4d1e634d1f196f45ce32f4c74e1bf60c7308650b2) => 1066748 (00000021020b2db01c41a4eacd15cdc032ab1d89c4e937b98f6d0d2e3a35b107)
2023-01-25 15:54:19.626 [TRC] CORE: Processing tip change for dcr
2023-01-25 15:54:19.752 [DBG] CORE[dcr]: Found mined tx 466fc2e455c4caf6c77046c21b98f82a036f5f516a18ec32ad99b655783f9246 in block 00000000f5faa35cce64914ac00788d700d71de051d218c8502461052780399e.
2023-01-25 15:54:19.752 [DBG] CORE[dcr]: Output 466fc2e455c4caf6c77046c21b98f82a036f5f516a18ec32ad99b655783f9246:0 is NOT spent in blocks 1066739 (000000009953b719bc823c350ef14452326757a64d597835773a66a5deeea2fd) to 1066748 (00000021020b2db01c41a4eacd15cdc032ab1d89c4e937b98f6d0d2e3a35b107).
```

----------

Restarting `dexc` seems to have resolved the issue (it results in re-scanning from earliest possible block that might contain Maker redemption), relevant logs after restart:
```
2023-01-25 18:29:58.624 [INF] CORE: Swap contract for match d76b65536370ca2e652b14d674eb0ddeb93b7950681d689cddef6d3069501565, order 35cd60a940b5aac4ad59893b0ce42416713e4b32da4c4c6019a9c422272d8245 is spent. Search for counterparty redemption may begin soon.
2023-01-25 18:29:58.624 [DBG] CORE: Ready to find counter-party redemption for match d76b65536370ca2e652b14d674eb0ddeb93b7950681d689cddef6d3069501565, order 35cd60a940b5aac4ad59893b0ce42416713e4b32da4c4c6019a9c422272d8245 (Taker)
2023-01-25 18:29:58.624 [DBG] CORE[dcr]: finding redemptions for 1 contracts in blocks 1066738 - 1066796
2023-01-25 18:29:58.625 [INF] CORE[dcr]: Redemption for contract 466fc2e455c4caf6c77046c21b98f82a036f5f516a18ec32ad99b655783f9246:0 found in tx input 57c6505bc9e3342a1b8ea199b978e52b2e06bde89dab0d8d0f51de51f134375c:0 in block 1066748
2023-01-25 18:29:58.625 [DBG] CORE[dcr]: 1 redemptions found, 0 canceled out of 1 contracts in blocks 1066738 to 1066748
2023-01-25 18:29:58.663 [DBG] CORE: Slow tick: trade 35cd60a940b5aac4ad59893b0ce42416713e4b32da4c4c6019a9c422272d8245 processed in 912.328413ms, blocked for 500ns
2023-01-25 18:29:58.702 [INF] CORE: Found redemption of contract 466fc2e455c4caf6c77046c21b98f82a036f5f516a18ec32ad99b655783f9246:0 (dcr) for order 35cd60a940b5aac4ad59893b0ce42416713e4b32da4c4c6019a9c422272d8245, match d76b65536370ca2e652b14d674eb0ddeb93b7950681d689cddef6d3069501565. Redeem: 57c6505bc9e3342a1b8ea199b978e52b2e06bde89dab0d8d0f51de51f134375c:0
```
and in UI also looking good:

<img width="200" alt="image" src="https://user-images.githubusercontent.com/112318969/214652593-67d1b026-f821-4a4a-bb4e-dd9e13803e79.png">

**Comment by chappjc** [comment]
There's logic in the "externaltx" code to check the mainchain status of the best block, but there might be a bug there.  @itswisdomagain anything come to mind?
This may also relate to the bug being fixed in https://github.com/decred/dcrdex/pull/2049/files#diff-4ff5cbeb6a9b212c78c955d636b154ea839b1c730b8ded6809c2c346c09267bcR331
Maybe we pull those two lines out into a separate commit for testing.

**Comment by norwnd** [comment]
Attaching logs before/after restart, in case you wanna cross-check my reasoning:

[dexc-log-before-restart.txt](https://github.com/decred/dcrdex/files/10506244/dexc-log-before-restart.txt)
[dexc-log-after-restart.txt](https://github.com/decred/dcrdex/files/10506245/dexc-log-after-restart.txt)

**Comment by JoeGruffins** [comment]
I don't think this is related to the bug in fixed in #2049 as `findRedemptionsInBlockRange` does not look to use `isOutputSpent`

I would guess that it is due to a reorg. It searched up to 1066748 then that block was reorg'd off and we saw the block 1066749 next and didn't re-search 1066748. So, the tx was not in the one we search but showed up in the mainchain 1066748. This has got to be super lucky to hit though.

**Comment by norwnd** [comment]
> I would guess that it is due to a reorg. It searched up to 1066748 then that block was reorg'd off and we saw the block 1066749 next and didn't re-search 1066748. So, the tx was not in the one we search but showed up in the mainchain 1066748. This has got to be super lucky to hit though.

This is exactly what I meant :) and I hit it today again on `testnet`(it seems there are a lot of reorgs happening there - https://testnet.dcrdata.org/side - and it doesn't even display all the reorged blocks, didn't display "my version = 00000021020b2db01c41a4eacd15cdc032ab1d89c4e937b98f6d0d2e3a35b107" of 1066748; looks like **dcrdata** isn't aware of this hash), btw.

**Comment by JoeGruffins** [comment]
> looks like dcrdata isn't aware of this hash), btw.

It looks like it stores the first block of a "side chain" but not those after that belong to that side chain. It may not even have seen the block before it was forked off.

**Comment by chappjc** [comment]
It was just never on that fork, so it wasn't part of that reorg.

**Comment by JoeGruffins** [comment]
It does have the 1066747 before that one https://testnet.dcrdata.org/block/00000000a96f16971b1a09f4d1e634d1f196f45ce32f4c74e1bf60c7308650b2

**Comment by JoeGruffins** [comment]
Well, this is interesting, he is on the wrong chain until the logs end. It looks like the tx was just never mined from his node's point of view. Logs go up to 1066788 which does not exist https://testnet.dcrdata.org/search?search=000000b4fe1e3c609af8af4e26f638f830b0884e70c460e1b39c4997c4f88dc1

**Comment by JoeGruffins** [comment]
@norwnd It looks like the node also starts up looking at the wrong chain, but does find the redemption. The log for after starting up ends before there is another tip change. If you still have the logs, can you see if there is a line like `reorg detected from height` in there?

It doesn't look like there was a reorg to detect in the first set of logs.

**Comment by JoeGruffins** [comment]
The only tip change in the start up is 
```
2023-01-25 18:29:57.750 [DBG] CORE[dcr]: tip change: 1066788 (000000b4fe1e3c609af8af4e26f638f830b0884e70c460e1b39c4997c4f88dc1) => 1066796 (000000f967e466738386324341038d031f9b86b96738edbffbe20749defd4124)
```
And that last block is not on main chain currently, it could have been from his node's point of view.

But then:
```
2023-01-25 18:29:58.624 [DBG] CORE[dcr]: finding redemptions for 1 contracts in blocks 1066738 - 1066796
2023-01-25 18:29:58.625 [INF] CORE[dcr]: Redemption for contract 466fc2e455c4caf6c77046c21b98f82a036f5f516a18ec32ad99b655783f9246:0 found in tx input 57c6505bc9e3342a1b8ea199b978e52b2e06bde89dab0d8d0f51de51f134375c:0 in block 1066748
2023-01-25 18:29:58.625 [DBG] CORE[dcr]: 1 redemptions found, 0 canceled out of 1 contracts in blocks 1066738 to 1066748
```

So, why didn't we detect a reorg yet I wonder, and do we detect it the next tip change.

**Comment by norwnd** [comment]
> The log for after starting up ends before there is another tip change. If you still have the logs, can you see if there is a line like reorg detected from height in there?

Don't have those logs anymore, for **after** log attached [attached above](https://github.com/decred/dcrdex/issues/2076#issuecomment-1404543299) I thought there is no reason to keep logging after this line:
```
2023-01-25 18:29:58.702 [INF] CORE: Found redemption of contract 466fc2e455c4caf6c77046c21b98f82a036f5f516a18ec32ad99b655783f9246:0 (dcr) for order 35cd60a940b5aac4ad59893b0ce42416713e4b32da4c4c6019a9c422272d8245, match d76b65536370ca2e652b14d674eb0ddeb93b7950681d689cddef6d3069501565. Redeem: 57c6505bc9e3342a1b8ea199b978e52b2e06bde89dab0d8d0f51de51f134375c:0
```

but I'll keep an eye out on `reorg detected from height` in case I encounter this issue  again.

**Comment by norwnd** [comment]
It does seem like my DCR SPV wallet was following a fork for a while. Why reorg wasn't detected for so long is kinda strange, I found only one place I think might be a possible culprit - here we check confirmations against `-1` to decide whether we can treat block as mainchain or not:

https://github.com/decred/dcrdex/blob/d82cd09329bbce716f50e21232172eeb07a6b644/client/asset/dcr/dcr.go#L4800-L4803

So, if rpc/spv wallet doesn't return `-1` for orphaned blocks there ^ we'll treat prev block we are checking in `handleTipChange` as mainchain and never try to unroll to mainchain ancestor. So the question is do all wallet implementation (rpc/spv) respect this `confirmations == -1` or not (that would explain that blind fork following). For RPC it's all in **dcrwallet** repo (hard to follow) and for for SPV `spvWallet.GetBlockHeader` calculates `Confirmations` in such a way that they are always `>= 1` ?

https://github.com/decred/dcrdex/blob/d82cd09329bbce716f50e21232172eeb07a6b644/client/asset/dcr/spv.go#L606-606

if that's correct assessment ^ we might want to check all types (RPC/SPV/node) of wallets for other coins too.

**Another** thing to note is **cfilters**, we use these only once (searching with `SwapConfirmations`) and then rely on cached value (that gets init/reset based on the rules above that are seemingly buggy). I haven't looked into how **cfilters** work, just putting it out here for future reference. That caching logic seems fine to me.

**Comment by chappjc** [comment]
> For RPC it's all in **dcrwallet** repo (hard to follow)

When dcrwallet is in RPC mode (not usespv), it is just a passthrough to dcrd.  When in spv mode, dcrwallet does this, which is essentially what dcrd does too: https://github.com/decred/dcrwallet/blob/39ec467c8000966099c138770640d779f93c54c1/internal/rpc/jsonrpc/methods.go#L1292-L1309
But you can just use dcrctl to see that.  So, yes, when using the dcrwallet app, -1 is returned for orphaned blocks.  This convention is the same with bitcoind and derivatives.

The issue is clearly `spvWallet`.  You've started some refactoring with `mainchainAncestor` it seems, so it looks like you're clear on how to fix it?

**Comment by chappjc** [comment]
`BlockInMainChain(ctx context.Context, hash *chainhash.Hash) (haveBlock, invalidated bool, err error)` is already defined as a method of the `dcrWallet interface`, so should be straightforward to resolve

**Comment by norwnd** [comment]
> You've started some refactoring with mainchainAncestor it seems, so it looks like you're clear on how to fix it?

Yes, shall I put that refactoring for review while we are at it ?
